### ISSUE TITLE
Perl 5.8.0 seg faults on exiting a do/while loop containing goto's

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6215]
Migrated from [rt.perl.org#20154](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=20154) (status was 'resolved')

Searchable as RT20154$

**Comment by p5pRT** [comment]
### From jon@trinity.fluff.org

#### Created by jon@trinity.fluff.org



The following code causes a seg fault and core dump on exit of the loop\.

my $i = 0;
\{
&nbsp;    do \{
&nbsp;        goto FOO;
&nbsp;        FOO&#8203;:
&nbsp;    \} while \($i \< \-1\);
\}

Removing the outer set of braces stops this from happening\, as does
removing the goto\. Using a variable in the package namespace seg faults
as well\.

A backtrace of the core dump in gdb shows the following&#8203;:
\(gdb\) bt
\#0  0x080bb702 in Perl\_sv\_free \(\)
\#1  0x080d9a19 in Perl\_pp\_leaveloop \(\)
\#2  0x080ae2e1 in Perl\_runops\_standard \(\)
\#3  0x0806202d in perl\_run \(\)
\#4  0x08061d3b in perl\_run \(\)
\#5  0x0805ee89 in main \(\)
\#6  0x400aba5f in \_\_libc\_start\_main \(\) from /lib/libc\.so\.6

This is running on a Debian testing/unstable machine under Linux\.

The above code doesn't cause a seg fault under Perl 5\.6\.0 or 5\.6\.1

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=critical

Site configuration information for perl v5.8.0:

Configured by Debian Project at Thu Nov 21 01:17:18 UTC 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.19, archname=i386-linux-thread-multi
    uname='linux cyberhq 2.4.19 #1 smp sun aug 4 11:30:45 pdt 2002 i686 pentium ii (deschutes) genuineintel gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8.0 -Darchlib=/usr/lib/perl/5.8.0 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.0 -Dsitearch=/usr/local/lib/perl/5.8.0 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.0 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20011002 (Debian prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.1.so, so=so, useshrplib=true, libperl=libperl.so.5.8.0
    gnulibc_version='2.3.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /sw/lib/site_perl
    /site/share/site_perl
    /etc/perl
    /usr/local/lib/perl/5.8.0
    /usr/local/share/perl/5.8.0
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8.0
    /usr/share/perl/5.8.0
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.0:
    HOME=/homes/jon
    LANG=en_GB
    LANGUAGE (unset)
    LC_ADDRESS=POSIX
    LC_ALL=POSIX
    LC_COLLATE=POSIX
    LC_CTYPE=POSIX
    LC_IDENTIFICATION=POSIX
    LC_MEASUREMENT=POSIX
    LC_MESSAGES=POSIX
    LC_MONETARY=POSIX
    LC_NAME=POSIX
    LC_NUMERIC=POSIX
    LC_PAPER=POSIX
    LC_TELEPHONE=POSIX
    LC_TIME=POSIX
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/conf/bin:/usr/sbin:/sbin:/usr/bin/X11:/usr/games:/sbin:/usr/sbin:/usr/local/sbin:/sw/opt/intel/compiler50/ia32/bin:/sw/bin:/usr/local/Acrobat4/bin
    PERL5LIB=/sw/lib/site_perl:/site/share/site_perl
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

This bug has been resolved by change \#18703\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
